ql/instruments/zerocouponswap.cpp

```diff
@@ -0,0 +1,147 @@
+            }
+            return floatCpn;
+        }
+    }
+
+    ZeroCouponSwap::ZeroCouponSwap(Type type,
+                                   Real baseNominal,
+                                   const Date& startDate,
+                                   const Date& maturityDate,
+                                   Real fixedPayment,
+                                   ext::shared_ptr<IborIndex> iborIndex,
+                                   const Calendar& paymentCalendar,
+                                   BusinessDayConvention paymentConvention,
+                                   Natural paymentDelay,
+                                   RateAveraging::Type averagingMethod)
+    : Swap(2), type_(type), baseNominal_(baseNominal), 
+      fixedPayment_(fixedPayment), iborIndex_(std::move(iborIndex)),
+      averagingMethod_(averagingMethod) {
+
+        QL_REQUIRE(!(baseNominal < 0.0), "base nominal cannot be negative");
+        QL_REQUIRE(startDate < maturityDate,
+                   "start date (" << startDate 
+                   << ") later than or equal to maturity date ("
+                   << maturityDate << ")");
+
+        Date paymentDate = paymentCalendar.advance(maturityDate, paymentDelay, 
+                                                   Days, paymentConvention);
+
+        legs_[0].push_back(
+            ext::shared_ptr<CashFlow>(new SimpleCashFlow(fixedPayment_, paymentDate)));
+        legs_[1].push_back(createSubPeriodicCoupon(paymentDate, startDate, maturityDate,
+                                                   baseNominal_, iborIndex_, averagingMethod_));
+        for (Leg::const_iterator i = legs_[1].begin(); i < legs_[1].end(); ++i)
+            registerWith(*i);
+
+        switch (type_) {
+            case Payer:
+                payer_[0] = -1.0;
+                payer_[1] = +1.0;
+                break;
+            case Receiver:
+                payer_[0] = +1.0;
+                payer_[1] = -1.0;
+                break;
+            default:
+                QL_FAIL("unknown zero coupon swap type");
+        }
+    }
+
+    ZeroCouponSwap::ZeroCouponSwap(Type type,
+                                   Real baseNominal,
+                                   const Date& startDate,
+                                   const Date& maturityDate,
+                                   Real fixedRate,
+                                   const DayCounter& fixedDayCounter,
+                                   ext::shared_ptr<IborIndex> iborIndex,
+                                   const Calendar& paymentCalendar,
+                                   BusinessDayConvention paymentConvention,
+                                   Natural paymentDelay,
+                                   RateAveraging::Type averagingMethod)
+    : ZeroCouponSwap(type, baseNominal, startDate, maturityDate,
+      calculateFixedPayment(startDate, maturityDate, baseNominal, fixedRate, fixedDayCounter),
```

Comment:
In the case of a fixed rate, it would be nice to have a coupon with rate information instead of just a `SimpleCashFlow`.  It would be useful for inflation swaps as well.  What do you think?

Reply:
Thank you for your review, Luigi. I quite like the idea to create a coupon in that case. I just pushed some changes to reflect that. Let me know if such setup is ok.

Reply:
Looks ok to me.